pkg/controller/logstash/sset/sset.go

```diff
@@ -0,0 +1,97 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package sset
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appsv1 "k8s.io/api/apps/v1"
+
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/hash"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/controller/common/reconciler"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/utils/k8s"
+	"github.com/elastic/cloud-on-k8s/v2/pkg/utils/maps"
+)
+
+type Params struct {
+	Name                 string
+	Namespace            string
+	ServiceName          string
+	Selector             map[string]string
+	Labels               map[string]string
+	PodTemplateSpec      corev1.PodTemplateSpec
+	Replicas             int32
+	RevisionHistoryLimit *int32
+}
+
+func New(params Params) (appsv1.StatefulSet, error) {
+	sset := appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      params.Name,
+			Namespace: params.Namespace,
+			Labels:    params.Labels,
+		},
+		Spec: appsv1.StatefulSetSpec{
+			UpdateStrategy: appsv1.StatefulSetUpdateStrategy{
+				Type: appsv1.RollingUpdateStatefulSetStrategyType,
+			},
+			// we don't care much about pods creation ordering, and manage deletion ordering ourselves,
+			// so we're fine with the StatefulSet controller spawning all pods in parallel
+			PodManagementPolicy:  appsv1.ParallelPodManagement,
+			RevisionHistoryLimit: params.RevisionHistoryLimit,
+			// build a headless service per StatefulSet, matching the StatefulSet labels
+			ServiceName: params.ServiceName,
```

Comment:
The service referenced here is not a headless service but the HTTP service for Logstash which has a ClusterIP. This means the individual Logstash Pods are not resolvable via DNS (because that service always resolves to the ClusterIP of the service and the routing then happens with the `iptables` magic of kube proxy). 

The K8s docs ask for a headless service for each stateful set. The thing I am not sure about is whether we need it if we don't need the ability to resolve the individual Logstash Pods via DNS.

Reply:
I changed it to headless service following the k8s doc. The default service in Logstash is only for metrics collection. Having load balancing with iptables sounds more than it needs.